### ISSUE TITLE
Docker daemon hangs intermittently after upgrading to docker 1.12.2

### ISSUE THREAD

**Issue by sakserv** [docker/docker#28754]
**Description**

We are encountering intermittent hangs of the docker daemon after moving to docker 1.12.2. Restarting docker appears to be the only way to recover. This appears to be similar to https://github.com/docker/docker/issues/25321.

It was also suggested that we move to docker 1.12.3. As this will require significant effort on our end, it would be great to get feedback on if the stack trace is indicative of an issue that was fixed between 1.12.2 and 1.12.3.

**Steps to reproduce the issue:**
1. Issue standard docker commands, happens intermittently across a large number of nodes, so I don't have a repeatable set to steps.

**Describe the results you received:**
Docker hangs intermittently. Restarting docker fixes the issue.

**Describe the results you expected:**
Docker shouldn't hang.

**Additional information you deem important (e.g. issue happens only occasionally):**
Only happening intermittently.

One potential workaround mentioned on https://github.com/docker/docker/issues/25321 is moving to overlayfs. We are using an lvm thinpool devicemapper setup leveraging xfs and unfortunately overlay is not very stable on Centos 7.2 so making that move is not currently an option. Doing so would be quite disruptive to existing workloads as well.

**Output of `docker version`:**

```
Client:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 105
 Running: 19
 Paused: 0
 Stopped: 86
Images: 108
Server Version: 1.12.2
Storage Driver: devicemapper
 Pool Name: vg01-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 274.9 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 853.2 GB
 Data Space Total: 5.63 TB
 Data Space Available: 4.777 TB
 Metadata Space Used: 105.9 MB
 Metadata Space Total: 16.98 GB
 Metadata Space Available: 16.87 GB
 Thin Pool Minimum Free Space: 563 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2015-12-01)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null overlay bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.13.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 251.6 GiB
Name: foo.example.com
ID: OSYP:WEPA:N2LF:KFTJ:BZNP:L3PT:LDNV:4OBJ:A4AM:CWFB:HHIK:WN4M
Docker Root Dir: /grid/0/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

Here is the stack trace from dockerd:
https://gist.github.com/sakserv/6aa7c7a1a8eac147e27d8c060023a36d

Docker client strace:
```
-snip-
etpeername(4, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, [23]) = 0
futex(0xc82004a908, FUTEX_WAKE, 1)      = 1
read(4, 0xc820327000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
write(4, "GET /v1.24/info HTTP/1.1\r\nHost: "..., 84) = 84
epoll_wait(5, {}, 128, 0)               = 0
futex(0x132cca8, FUTEX_WAIT, 0, NULL
```

dockerd strace:
```
Process 850182 attached
wait4(850188,
```

docker-containerd (pid 850188) strace:
```
Process 850188 attached
futex(0xee44c8, FUTEX_WAIT, 0, NULL
```

**Comment by grzr** [comment]
I did upgrade for docker from version 1.9.1-trusty to 1.12.3-trusty and I got same issue. Whats more none of docker commands worked (they just were hanging on "sth"). Never then less I was able to fix it. 

**TLDR;** 

- remove debugging flag "-D" from DOCKER_OPTS in `/etc/defaults/docker` and use `--debug` instead
- "-D" argument stopped to hanging docker commands after removing content of logs `/var/log/upstart/docker.log`


I don't know which combo of the steps helped while I was struggling with the issue so this why I will give here all my steps:

Update docker-engine
- `apt-get install docker-engine=1.12.3-0~trusty`
I've selected to use configuration from maintainer. Then I've put old variables
```
DOCKER_OPTS='-D --insecure-registry company-registry01 --insecure-registry company-registry02'
export HTTP_PROXY="http://company-proxy/"
export HTTPS_PROXY=Https Proxy: https://company-proxy/
export NO_PROXY=, company-docker,/var/run/docker.sock,/var/run/docker/libcontainerd/docker-containerd.sock
```
- `service docker restart`
- Start of realisation that docker commands are hanging: `docker ps`
- is dockerd there?
```
ps aux | grep docker
root      8431  0.4  0.5 1198976 39084 ?       Ssl  10:46   1:29 /usr/bin/dockerd --insecure-registry company-registry01 --insecure-registry company-regsitry02 --raw-logs
root      8443  0.0  0.1 514276  6948 ?        Ssl  10:46   0:01 docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --shim docker-containerd-shim --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/containerd --runtime docker-rune
```
- I've checked the Internet to found this and all preceding  issues with the topic "docker command hanging", no concrete solution found
- AppArmor? no entries found for `aa-logprof`
- lets try again 
```
apt-get remove docker-engine 
apt-get install docker-engine
```
- for certainty I've did
```
service apparmor restart
service docker restart
```
- ps showed to me that docker service was still running
`
kill -9 proc1 proc2
`
- after `service docker start` still the commands were hanging
- I've removed from `/etc/default/docker` variable DOCKER_OPTS
- FIXED
- I've tried to put whole DOCKER_OPTS and found out that issue was the short named argument `-D`, it was all OK with `-debug`
- after `truncate -s0 /var/log/upstart/docker.log` the argument "-D" was working again

Output from working docker info:

/var/lib/docker is a separately mounted hard drive

```
Containers: 6
 Running: 3
 Paused: 0
 Stopped: 3
Images: 270
Server Version: 1.12.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 318
 Dirperm1 Supported: false
Logging Driver: son-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runs
Default Runtime: runs
Security Options: apparmor
Kernel Version: 3.13.0-91-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 6.418 GiB
Name: company-docker
ID: RVHF:AJTO:O5KP:TTR3:NN6T:CSO6:LOJ4:2VY2:2E64:WYIA:GYBX:AOJX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://company-proxy/
Https Proxy: https://company-proxy/
No Proxy: , company-host,/var/run/docker.sock,/var/run/docker/libcontainerd/docker-containerd.sock
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 company-registry01
 company-registry02
 127.0.0.0/8
```

Logs with timestamps are attached.


I hope that my short journey will help someone else with the issue.

[docker.txt](https://github.com/docker/docker/files/613644/docker.txt)

**Comment by thaJeztah** [comment]
@grzr from those logs it looks like docker wasn't actually hanging, but working on the migration to content addressable store; see the message at the top of the 1.10 changelog; https://github.com/docker/docker/releases/tag/v1.10.0

**Comment by grzr** [comment]
@thaJeztah I did update in the morning then tried again with flag -D afternoon. Last info-log entry about migration is:
```
time="2016-11-25T10:41:53.682897842+01:00" level=info msg="migrated tag bs:latest to point to sha256:6d9a221b79827213904c9a9d64f4c2935c50281e8d1121112d5c1ee3dee0af5d"
```
docker service wasn't working with flag -D/--debug:
- few minutes after last migration appeared: 10:45
- six hours after last migration entry: ~16:30

I regret that I didn't have it run in Ansible to be more helpful.

**Comment by sakserv** [comment]
As a follow up, we have not encountered a similar hang in over a month. We are rolling out an upgrade to docker and enabling debug. I'll add logs if this happens again.

**Comment by thaJeztah** [comment]
@sakserv let me close this issue for now, but let us know if you're still running into this

**Comment by sakserv** [comment]
We recently completed the upgrade to Docker 1.12.5. 

Prior to upgrade, this issue had returned, but was only impacting around 1 node per week. 

Post upgrade, Docker completely hung and needed to be restarted to recover on 13 out of ~200 nodes over the weekend alone, so it seems this issue may be worse post upgrade.

We have enabled debugging, and one interesting observation is that the API call to dockerd is still logged when running, say, {{docker ps}}. See the log entry below. (note that when the issue occurs, {{docker ps}} will never return unless dockerd is restarted). 

For example:

```
Feb 21 12:30:35 server.foo.com dockerd[761487]: time="2017-02-21T12:30:35.473102800Z" level=debug msg="Calling GET /v1.24/containers/json"
```

To me, this appears to indicate that the hang isn't dockerd itself, but communication down the line. 

Any thoughts on how we could troubleshoot this further?

**Comment by thaJeztah** [comment]
If a daemon is stuck, you can send a `SIGUSR1` to the `dockerd` process, which will dump a stack-trace in the daemon logs (docker 1.12) or in a separate file (docker 1.13). That dump should provide more information what is happening. It's still possible that there's a deadlock in the daemon; work is in progress on improving that, but design/approach is being discussed (see https://github.com/docker/docker/issues/30225)

**Comment by sakserv** [comment]
Thanks @thaJeztah - we just had a daemon hang. I have captured the stack trace here: https://gist.github.com/sakserv/9d179e115fbec5ae7dc672d4826a2ad0

Please let me know if anything stands out.

**Comment by sakserv** [comment]
In looking for go stack trace analyzers, I found a python tool that parses the stack trace output, groups by call type, and counts the created by instances for each. It seems there is a truncation issue with some function names, but posting it here in case it helps.

```
syscall (42):
   Created by:
	github.com/docker/docker/contain: 1
	github.com/docker/docker/container/stream.(*Config).CopyToPipe.func1: 38
	github.com/docker/docker/con: 1
	github.com/docker/docker/containe: 1
	github.com/docker/docker/distribution/xfer.(*LayerDownloadManager).makeDownloadFunc.func1: 1

IO wait (4):
   Created by:
	net/http.(*connReader).startBackgroundRead: 2
	github.com/docker/docker/api/server.(*Server).serveAPI: 1
	github.com/docker/libnetwork.(*controller).startExternalKeyListener: 1

chan receive (5):
   Created by:
	github.com/docker/docker/daemon.NewDaemon: 1
	main.(*DaemonCli).setupConfigReloadTrap: 1
	github.com/docker/docker/daemon.(*Daemon).pullImageWithReference: 2
	main.(*DaemonCli).start: 1

semacquire (56):
   Created by:
	github.com/docker/docker/daemon/logger.(*: 1
	github.com/docker/docker/daemon/logger.(*Copier).Run: 38
	net/http.(*Server).Serve: 12
	github.com/docker/docker/distribution/xfer.(*LayerDownloadManager).makeDownloadFunc.func1: 1
	github.com/vbatts/tar-split/tar/asm.NewInputTarStream: 1
	github.com/docker/docker/pkg/ioutils.NewCancelReadCloser: 1
	github.com/docker/docker/daemon/graphdriver/devmapper.(*DeviceSet).initMetaData: 1
	github.com/docker: 1

select (16):
   Created by:
	github.com/docker/docker/libcontainerd.(*remote).startEventsMonitor: 1
	google.golang.org/grpc/transport.newHTTP2Client: 1
	net/http.(*Server).Serve: 2
	github.com/docke: 1
	google.golang.org/grpc.NewClientStream: 1
	google.golang.org/grpc.(*Conn).WaitForStateChange: 1
	github.com/docker/docker/distribution/xfer.(*transferManager).Transfer: 3
	github.com/docker/docker/pkg/ioutils.NewCancelReadCloser: 1
	github.com/docker/docker/distribution/xfer.(*transfer).Watch: 2
	github.com/docker/docker/api/server/router.cancellableHandler.func1: 2
	github.com/docker/libnetwork.(*controller).startWatch: 1
```

**Comment by sakserv** [comment]
I just had a thought. We are now seeing more of these post upgrade, but we also enabled debug logging. Given all of the logger.Copier entries, are we perhaps filling some logging buffer that is resulting in the daemon blocking?

**Comment by thaJeztah** [comment]
Could be possible. Let me ping @mlaventure @cpuguy83 who are better at looking at the stack trace / possible causes 😅 .

**Comment by cpuguy83** [comment]
No, this looks like the devmapper driver is locked, waiting on a response from udev: `github.com/docker/docker/pkg/devicemapper._Cfunc_dm_udev_wait(0xc80d4da465, 0xc800000000)`

**Comment by thaJeztah** [comment]
Looks like it may be the same issue as https://github.com/docker/docker/issues/27900 then?

**Comment by sakserv** [comment]
Thanks @cpuguy83 and @thaJeztah - device mapper strikes again - it seems dm + non-trivial load is simply a recipe for disaster. Given our limited options, we'll try increasing the udevd event timeout as discussed in https://github.com/docker/docker/pull/29029.

**Comment by thaJeztah** [comment]
Overlay support on RHEL/CentOS kernels is also picking up, and (I think) overlay2 should even be working on their recent kernels.

**Comment by coolljt0725** [comment]
@sakserv 1. once it hangs, use `dmsetup udevcomplete_all` to recover. 2. increase the `--event-timeout` does help. 3. you also can switch to `overlay` if your kernel support. 
From our analysis, if there are some containers doing some heavy IO, it may cause this issue.
